portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java

```diff
@@ -223,6 +236,134 @@ public void testStripSuffixWhenFileNameHasParenthesisAtStart() {
 			"()test.jsp", _fileImpl.stripParentheticalSuffix("()test.jsp"));
 	}
 
+	@Test
+	public void testUnzip() throws Exception {
+		_setUpForUnzipTests();
+
+		File file = _createZipFile(
+			"test.zip",
+			HashMapBuilder.put(
+				"zip/test/directory/", () -> (String)null
+			).put(
+				"zip/test/entry/entry.txt", "Test String"
+			).build());
+
+		_fileImpl.unzip(file, _tempFolder);
+
+		_assertExists("zip/test/directory");
+		_assertExists("zip/test/entry/entry.txt");
+
+		Assert.assertTrue(
+			_canFileReadWrite(_getTempTestFilePath("zip/test/directory")));
+
+		Assert.assertTrue(
+			_canFileReadWrite(
+				_getTempTestFilePath("zip/test/entry/entry.txt")));
+		_tearDown();
+	}
+
+	@Test
+	public void testUnzipZipSlipVulnerable() throws Exception {
+		_setUpForUnzipTests();
+
+		File file = _createZipFile(
+			"test_slip.zip",
+			HashMapBuilder.put(
+				"../../../../../../bad.txt", "I am bad!"
+			).put(
+				"good.txt", "I am good!"
+			).build());
+
+		_fileImpl.unzip(file, _tempFolder);
+
+		_assertExists("good.txt");
+		_assertDoesNotExist("tmp/bad.txt");
```

Comment:
@jeffwu0724 I have a question about this assertion. The bad file has a path like `../../../../../../bad.txt`, but why is the assertion asserting `$_tempFolder/tmp/bad.txt`? It seems the zip content is just unzipped to the `_tempFolder`, so the bad file should be out of it, if the fix is not applied?

Could you double check this part, and make sure it fails without the `FileImpl` fix, while passes with the fix?

Reply:
Hey Dante, seems like if I don't have the fix, it will show 
`Testcase: testUnzipZipSlipVulnerable(com.liferay.portal.util.FileImplTest): Caused an ERROR
    [junit] /tmp/20220607215702662XSIATEBR/../../../../../../bad.txt
    [junit] java.nio.file.AccessDeniedException: /tmp/20220607215702662XSIATEBR/../../../../../../bad.txt
    [junit]     at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84)
    [junit]     at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
    [junit]     at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
    [junit]     at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
    [junit]     at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
    [junit]     at java.nio.file.Files.newOutputStream(Files.java:216)
    [junit]     at java.nio.file.Files.copy(Files.java:3016)
    [junit]     at com.liferay.portal.util.FileImpl.unzip(FileImpl.java:849)
    [junit]     at com.liferay.portal.util.FileImplTest.testUnzipZipSlipVulnerable(FileImplTest.java:263)
    [junit]     at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
    [junit]     at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:`
instead of just go to the _assertDoesNotExist("tmp/bad.txt");. I have changed my logic here according to the reference you sent me.

Reply:
@jeffwu0724 Yeah, here the error is because the Java process tries to extract this file to `/`, which is not writable unless you `sudo`. If you look at the fix and test made by Search Team, you'll see that they created the bad file to be `../../../../../../../../../../../../../tmp/bad.txt`. I guess this ensures the file to be extracted into `/tmp`.

However, `/tmp` only exists in *Unix. If you run it on Windows, I'm not sure about the behavior. I think the best way to test it is to create the file to be `../bad.txt` and extract it into a sub-dir under the temp dir we use. So without the fix, the file will be extracted into the temp dir and the test can easily find it.